server/routes/user.js

```diff
@@ -1,30 +1,139 @@
 const express = require("express");
-const { check, validationResult } = require("express-validator");
+const { check, body, validationResult } = require("express-validator");
 const router = express.Router();
+const bcrypt = require("bcryptjs");
+const jwt = require("jsonwebtoken");
 
 const { User } = require("../database");
 
 router.post(
-  "/register",
+  "/signup",
   [
     check("name", "Name is required")
       .not()
       .isEmpty(),
-    check("email", "Please use a valid email address").isEmail()
+    check("email", "Please use a valid email address").isEmail(),
+    check("password", "Choose a password at least 6 characters long").isLength({
+      min: 6
+    }),
+    body("password2").custom((value, { req }) => {
+      if (value !== req.body.password) {
+        throw new Error("Password confirmation does not match password");
+      }
+      return true;
+    })
   ],
-  async (req, res) => {
+  (req, res) => {
     const errors = validationResult(req);
-    if (errors.length) {
+    if (errors.length > 0) {
       return res.status(400).json({ errors: errors.array() });
     }
-    const { name, email } = req.body;
-    try {
-      const user = new User({ name, email });
-      await user.save();
-      res.send(user);
-    } catch (err) {
-      console.err(err.message);
+
+    const { name, email, password } = req.body;
+
+    //Check for an existing user
+    User.findOne({ email: email }).then(user => {
+      if (user) {
+        return res.status(400).json({ errors: ["Email already exists"] });
+      } else {
+        const newUser = new User({
+          name: name,
+          email: email,
+          password: password,
+          experience: {}
+        });
+
+        //hash the incoming password and save to DB
+        bcrypt.genSalt(10, (err, salt) => {
+          bcrypt.hash(newUser.password, salt, (err, hash) => {
+            if (err) throw err;
+            newUser.password = hash;
+            newUser
+              .save()
+              .then(user => {
+                //On successful save, log user in and send JW token
+                const payload = {
+                  email: user.email,
+                  name: user.name
+                };
+                jwt.sign(
+                  payload,
+                  process.env.SECRET,
+                  {
+                    expiresIn: 2628000 //1 month
```

Comment:
magic number here - should be a constant somewhere, maybe in a config file

Reply:
I updated the code to manage that variable from .env since that's where our other variables are. Do we want to use a different configuration file or is that okay?